File: src/format.ts
export interface WeeklySummary {
  weekStart: string; // YYYY-MM-DD
  weekEnd: string;   // YYYY-MM-DD
  clients: ClientSummary[];
  grandTotal: number;
}

export interface ClientSummary {
  name: string;
  daily: number[]; // hours for Mon–Fri (5 elements)
  total: number;
}

const DAY_HEADERS = ["Mon", "Tue", "Wed", "Thu", "Fri"];

function formatHours(h: number): string {
  if (h === 0) return "—";
  return h.toFixed(1);
}

function formatDateRange(start: string, end: string): string {
  const s = new Date(start + "T00:00:00");
  const e = new Date(end + "T00:00:00");
  const months = [
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
  ];
  return `${months[s.getMonth()]} ${s.getDate()} – ${months[e.getMonth()]} ${e.getDate()}, ${e.getFullYear()}`;
}

export function formatTable(summary: WeeklySummary): string {
  const COL_WIDTH = 6;
  const NAME_WIDTH = 20;
  const lines: string[] = [];

  lines.push(`Week of ${formatDateRange(summary.weekStart, summary.weekEnd)}`);
  lines.push("");

  // Header
  const header =
    "Client".padEnd(NAME_WIDTH) +
    DAY_HEADERS.map((d) => d.padStart(COL_WIDTH)).join("") +
    "  Total";
  lines.push(header);

  const separator = "─".repeat(header.length);
  lines.push(separator);

  // Client rows
  for (const client of summary.clients) {
    const row =
      client.name.slice(0, NAME_WIDTH).padEnd(NAME_WIDTH) +
      client.daily.map((h) => formatHours(h).padStart(COL_WIDTH)).join("") +
      formatHours(client.total).padStart(COL_WIDTH + 1) +
      "h";
    lines.push(row);
  }

  lines.push(separator);

  // Totals row
  const dailyTotals = [0, 0, 0, 0, 0];
  for (const client of summary.clients) {
    for (let i = 0; i < 5; i++) {
      dailyTotals[i]! += client.daily[i]!;
    }
  }
  const totalsRow =
    "Total".padEnd(NAME_WIDTH) +
    dailyTotals.map((h) => formatHours(h).padStart(COL_WIDTH)).join("") +
    formatHours(summary.grandTotal).padStart(COL_WIDTH + 1) +
    "h";
  lines.push(totalsRow);

  return lines.join("\n");
}

export function formatJson(summary: WeeklySummary): string {
  return JSON.stringify(summary, null, 2);
}

File: internal/format/format.go
package format

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// WeeklySummary holds the weekly time data.
type WeeklySummary struct {
	WeekStart  string          `json:"weekStart"`
	WeekEnd    string          `json:"weekEnd"`
	Clients    []ClientSummary `json:"clients"`
	GrandTotal float64         `json:"grandTotal"`
}

// ClientSummary holds per-client daily hours.
type ClientSummary struct {
	Name  string    `json:"name"`
	Daily []float64 `json:"daily"` // Mon–Fri (5 elements)
	Total float64   `json:"total"`
}

var dayHeaders = []string{"Mon", "Tue", "Wed", "Thu", "Fri"}

func formatHours(h float64) string {
	if h == 0 {
		return "—"
	}
	return fmt.Sprintf("%.1f", h)
}

func formatDateRange(start, end string) string {
	months := []string{"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}
	s, _ := time.Parse("2006-01-02", start)
	e, _ := time.Parse("2006-01-02", end)
	return fmt.Sprintf("%s %d – %s %d, %d", months[s.Month()-1], s.Day(), months[e.Month()-1], e.Day(), e.Year())
}

// Table renders a WeeklySummary as a formatted text table.
func Table(summary *WeeklySummary) string {
	const colWidth = 6
	const nameWidth = 20

	var lines []string

	lines = append(lines, fmt.Sprintf("Week of %s", formatDateRange(summary.WeekStart, summary.WeekEnd)))
	lines = append(lines, "")

	// Header
	header := fmt.Sprintf("%-*s", nameWidth, "Client")
	for _, d := range dayHeaders {
		header += fmt.Sprintf("%*s", colWidth, d)
	}
	header += "  Total"
	lines = append(lines, header)

	separator := strings.Repeat("─", len(header))
	lines = append(lines, separator)

	// Client rows
	for _, client := range summary.Clients {
		name := client.Name
		if len(name) > nameWidth {
			name = name[:nameWidth]
		}
		row := fmt.Sprintf("%-*s", nameWidth, name)
		for _, h := range client.Daily {
			row += fmt.Sprintf("%*s", colWidth, formatHours(h))
		}
		row += fmt.Sprintf("%*s", colWidth+1, formatHours(client.Total))
		row += "h"
		lines = append(lines, row)
	}

	lines = append(lines, separator)

	// Totals row
	dailyTotals := make([]float64, 5)
	for _, client := range summary.Clients {
		for i := 0; i < 5; i++ {
			dailyTotals[i] += client.Daily[i]
		}
	}
	totals := fmt.Sprintf("%-*s", nameWidth, "Total")
	for _, h := range dailyTotals {
		totals += fmt.Sprintf("%*s", colWidth, formatHours(h))
	}
	totals += fmt.Sprintf("%*s", colWidth+1, formatHours(summary.GrandTotal))
	totals += "h"
	lines = append(lines, totals)

	return strings.Join(lines, "\n")
}

// JSON renders a WeeklySummary as indented JSON.
func JSON(summary *WeeklySummary) string {
	data, _ := json.MarshalIndent(summary, "", "  ")
	return string(data)
}
